File: jlox/ts/class/Scanner.ts
import { Lox } from "./Lox.ts";
import { Token } from "./Token.ts";
import { TokenType } from "./TokenType.ts";

const keywords: {
    [key: string]: TokenType;
} = {
    and: TokenType.AND,
    class: TokenType.CLASS,
    else: TokenType.ELSE,
    false: TokenType.FALSE,
    for: TokenType.FOR,
    fun: TokenType.FUN,
    if: TokenType.IF,
    nil: TokenType.NIL,
    or: TokenType.OR,
    print: TokenType.PRINT,
    return: TokenType.RETURN,
    super: TokenType.SUPER,
    this: TokenType.THIS,
    true: TokenType.TRUE,
    var: TokenType.VAR,
    while: TokenType.WHILE,
};

export class Scanner {
    source: string;
    tokens: Token[];
    // Points to first character in the lexeme being scanned
    start = 0;
    // Points at the character currently being considered
    current = 0;
    line = 1;
    constructor(source: string) {
        this.source = source;
        this.tokens = [];
    }

    scanTokens() {
        while (!this.isAtEnd()) {
            this.start = this.current;
            this.scanToken();
        }
        this.tokens.push(
            new Token(
                TokenType.EOF,
                "",
                null,
                this.line,
            ),
        );
        return this.tokens;
    }

    scanToken() {
        const c = this.advance();
        switch (c) {
            // Single-character lexems
            case "(":
                this.addToken(TokenType.LEFT_PAREN);
                break;
            case ")":
                this.addToken(TokenType.RIGHT_PAREN);
                break;
            case "{":
                this.addToken(TokenType.LEFT_BRACE);
                break;
            case "}":
                this.addToken(TokenType.RIGHT_BRACE);
                break;
            case ",":
                this.addToken(TokenType.COMMA);
                break;
            case ".":
                this.addToken(TokenType.DOT);
                break;
            case "-":
                this.addToken(TokenType.MINUS);
                break;
            case "+":
                this.addToken(TokenType.PLUS);
                break;
            case ";":
                this.addToken(TokenType.SEMICOLON);
                break;
            case "*":
                this.addToken(TokenType.STAR);
                break;
            // Single to double character lexemes
            case "!":
                this.addToken(
                    this.match("=") ? TokenType.BANG_EQUAL : TokenType.BANG,
                );
                break;
            case "<":
                this.addToken(
                    this.match("=") ? TokenType.LESS_EQUAL : TokenType.LESS,
                );
                break;
            case ">":
                this.addToken(
                    this.match("=")
                        ? TokenType.GREATER_EQUAL
                        : TokenType.GREATER,
                );
                break;
            // Longer lexemes
            case "/":
                if (this.match("/")) {
                    // Comments go to end of line
                    while (this.peek() != "\n" && !this.isAtEnd()) {
                        this.advance();
                    }
                } else {
                    this.addToken(TokenType.SLASH);
                }
                break;
            // Newlines and whitespace
            case " ":
            case "\r":
            case "\t":
                // Ignore whitespace.
                break;
            case "\n":
                this.line++;
                break;
            // String literals
            case '"':
                this.scanString();
                break;

            default:
                if (this.isDigit(c)) {
                    this.scanNumber();
                } else if (this.isAlpha(c)) {
                    this.identifier();
                } else {
                    Lox.error(this.line, "Unexpected character.");
                }
                break;
        }
    }

    identifier() {
        while (this.isAlphaNumeric(this.peek())) {
            this.advance();
        }
        const text = this.source.slice(this.start, this.current);
        let type = keywords[text];
        if (type == null) {
            type = TokenType.IDENTIFIER;
        }
        this.addToken(TokenType.IDENTIFIER);
    }

    scanNumber() {
        while (this.isDigit(this.peek())) {
            this.advance();
        }

        // Look for a fractional part.
        if (this.peek() == "." && this.isDigit(this.peekNext())) {
            // Consume the "."
            this.advance();

            while (this.isDigit(this.peek())) {
                this.advance();
            }
        }

        this.pushToken(
            TokenType.NUMBER,
            Number.parseFloat(this.source.substring(this.start, this.current)),
        );
    }

    scanString() {
        while (this.peek() != '"' && !this.isAtEnd()) {
            if (this.peek() == "\n") {
                this.line++;
            }
            this.advance();
        }

        if (this.isAtEnd()) {
            Lox.error(this.line, "Unterminated string");
            return;
        }

        // The closing "
        this.advance();

        // Trim the surrounding quotes
        const value = this.source.slice(this.start + 1, this.current - 1);
        this.pushToken(TokenType.STRING, value);
    }

    isAlpha(c: string): boolean {
        const code = c.charCodeAt(0);
        if (
            (64 < code && code < 91) || // upper alpha (A-Z)
            (96 < code && code < 123) || // lower alpha (a-z)
            (code == 95) // underscore _
        ) {
            return true;
        }
        return false;
    }

    isDigit(c: string) {
        const code = c.charCodeAt(0);
        if (47 < code && 58 > code) {
            return true;
        }
        return false;
        // return !Number.isNaN(c);
    }

    isAlphaNumeric(c: string) {
        return this.isAlpha(c) || this.isDigit(c);
    }

    advance(): string {
        return this.source.charAt(this.current++);
    }

    peek() {
        if (this.isAtEnd()) {
            return "\0";
        }
        return this.source.charAt(this.current);
    }

    peekNext(): string {
        if (this.current + 1 >= this.source.length) {
            return "\0";
        }
        return this.source.charAt(this.current + 1);
    }

    match(expected: string): boolean {
        if (this.isAtEnd()) {
            return false;
        }
        if (this.source.charAt(this.current) != expected) {
            return false;
        }
        this.current++;
        return true;
    }

    addToken(type: TokenType) {
        this.pushToken(type, null);
    }

    pushToken(type: TokenType, literal: any) {
        const text = this.source.slice(this.start, this.current);
        this.tokens.push(
            new Token(
                type,
                text,
                literal,
                this.line,
            ),
        );
    }
    isAtEnd(): boolean {
        return this.source.length <= this.current;
    }
}

File: jlox/rust/src/book/scanner.rs
use super::tokens::{Token, TokenType};

pub struct Scanner {
    source: String,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
}

impl Scanner {
    pub fn from(source: String) -> Scanner {
        Scanner {
            source,
            tokens: Vec::new(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    pub fn scan_tokens(&mut self) -> &Vec<Token> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token();
        }
        let lexeme = self.source.get(self.start..self.current).unwrap();
        self.tokens.push(Token {
            token_type: TokenType::Eof,
            lexeme: lexeme.to_string(),
            line: self.line,
        });
        &self.tokens
    }

    fn is_at_end(&self) -> bool {
        return self.current >= self.source.len();
    }

    fn scan_token(&mut self) {
        let c = self.advance();
        match c {
            '(' => self.add_token(TokenType::ParenOpen),
            ')' => self.add_token(TokenType::ParenClosed),
            '{' => self.add_token(TokenType::BraceOpen),
            '}' => self.add_token(TokenType::BraceClosed),
            ',' => self.add_token(TokenType::Comma),
            ';' => self.add_token(TokenType::Semicolon),
            '.' => self.add_token(TokenType::Dot),
            '-' => self.add_token(TokenType::Minus),
            '+' => self.add_token(TokenType::Plus),
            '*' => self.add_token(TokenType::Star),
            '!' => {
                if self.match_char('=') {
                    self.add_token(TokenType::BangEq);
                } else {
                    self.add_token(TokenType::Bang);
                }
            }
            '=' => {
                if self.match_char('=') {
                    self.add_token(TokenType::EqEq);
                } else {
                    self.add_token(TokenType::Eq);
                }
            }
            '<' => {
                if self.match_char('=') {
                    self.add_token(TokenType::LessEq);
                } else {
                    self.add_token(TokenType::Less);
                }
            }
            '>' => {
                if self.match_char('=') {
                    self.add_token(TokenType::GreaterEq);
                } else {
                    self.add_token(TokenType::Greater);
                }
            }
            '/' => {
                match self.peek() {
                    '/' => {
                        // A comment goes until the end of the line.
                        while self.peek() != '\n' && !self.is_at_end() {
                            self.advance();
                        }
                    }
                    '*' => {
                        // A block comment goes until the end of the block.
                        while self.peek() != '*' && self.peek_next() != '/' && !self.is_at_end() {
                            if self.peek() == '\n' {
                                self.line += 1;
                            }
                            self.advance();
                        }
                        // Consume the closing */
                        self.advance();
                        self.advance();
                    }
                    _ => self.add_token(TokenType::Slash),
                    
                }
            }
            ' ' | '\r' | '\t' => (),
            '\n' => self.line += 1,
            '"' => self.string(),
            _ => {
                if c.is_ascii_digit() {
                    self.number();
                } else if c.is_ascii_alphabetic() {
                    self.identifier();
                } else {
                    panic!("Unexpected character: {}", c);
                }
            }
        }
    }

    fn advance(&mut self) -> char {
        self.current += 1;
        self.source.chars().nth(self.current - 1).unwrap()
    }

    fn add_token(&mut self, token_type: TokenType) {
        let lexeme = self.source.get(self.start..self.current).unwrap();
        self.tokens.push(Token {
            token_type,
            lexeme: lexeme.to_string(),
            line: self.line,
        });
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }
        if self.source.chars().nth(self.current).unwrap() != expected {
            return false;
        }
        self.current += 1;
        true
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            return '\0';
        }
        self.source.chars().nth(self.current).unwrap()
    }

    fn string(&mut self) {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            panic!("Unterminated string.");
        }

        // The closing "
        self.advance();

        // Trim the surrounding quotes
        let lexeme = self.source.get(self.start + 1..self.current - 1).unwrap();
        self.tokens.push(Token {
            token_type: TokenType::String(lexeme.to_string()),
            lexeme: lexeme.to_string(),
            line: self.line,
        });
    }

    fn number(&mut self) {
        while self.peek().is_ascii_digit() {
            self.advance();
        }

        // Look for a fractional part.
        if self.peek() == '.' && self.peek_next().is_ascii_digit() {
            // Consume the "."
            self.advance();

            while self.peek().is_ascii_digit() {
                self.advance();
            }
        }

        let lexeme = self.source.get(self.start..self.current).unwrap();
        self.tokens.push(Token {
            token_type: TokenType::Number(lexeme.to_string()),
            lexeme: lexeme.to_string(),
            line: self.line,
        });
    }

    fn peek_next(&self) -> char {
        if self.current + 1 >= self.source.len() {
            return '\0';
        }
        self.source.chars().nth(self.current + 1).unwrap()
    }

    fn identifier(&mut self) {
        while self.peek().is_ascii_alphanumeric() {
            self.advance();
        }

        let lexeme = self.source.get(self.start..self.current).unwrap();
        let token_type = match lexeme {
            "class" => TokenType::Class,
            "else" => TokenType::Else,
            "false" => TokenType::False,
            "for" => TokenType::For,
            "if" => TokenType::If,
            "nil" => TokenType::Nil,
            "return" => TokenType::Return,
            "super" => TokenType::Super,
            "this" => TokenType::This,
            "true" => TokenType::True,
            "var" => TokenType::Var,
            "while" => TokenType::While,
            _ => TokenType::Identifier(lexeme.to_string()),
        };
        self.tokens.push(Token {
            token_type,
            lexeme: lexeme.to_string(),
            line: self.line,
        });
    }
}
